src/include/test/unit_scoring.cc

```diff
@@ -36,9 +36,91 @@
+    CHECK(sum_of_squares(c, a) == 10);
+    CHECK(sum_of_squares(a, d) == 10);
+    CHECK(sum_of_squares(d, a) == 10);
+    CHECK(sum_of_squares(a, z) == 985);
+  }
+
+  SECTION("sum_of_squares_distance function object") {
+    CHECK(sum_of_squares_distance{}(a, b) == 0);
+    CHECK(sum_of_squares_distance{}(a, c) == 10);
+    CHECK(sum_of_squares_distance{}(c, a) == 10);
+    CHECK(sum_of_squares_distance{}(a, d) == 10);
+    CHECK(sum_of_squares_distance{}(d, a) == 10);
+    CHECK(sum_of_squares_distance{}(a, z) == 985);
+  }
+  SECTION("sub_sum_of_squares function") {
+    CHECK(sub_sum_of_squares(a, b, 0, size(a)) == 0);
+    CHECK(sub_sum_of_squares(a, c, 0, size(a)) == 10);
+    CHECK(sub_sum_of_squares(c, a, 0, size(a)) == 10);
+    CHECK(sub_sum_of_squares(a, d, 0, size(a)) == 10);
+    CHECK(sub_sum_of_squares(d, a, 0, size(a)) == 10);
+    CHECK(sub_sum_of_squares(a, z, 0, size(a)) == 985);
+
+    CHECK(sub_sum_of_squares(a, b, 0, size(a) / 2) == 0);
+    CHECK(sub_sum_of_squares(a, c, 0, size(a) / 2) == 5);
+    CHECK(sub_sum_of_squares(c, a, 0, size(a) / 2) == 5);
+    CHECK(sub_sum_of_squares(a, d, 0, size(a) / 2) == 5);
+    CHECK(sub_sum_of_squares(d, a, 0, size(a) / 2) == 5);
+    CHECK(sub_sum_of_squares(a, z, 0, size(a) / 2) == 255);
+
+    CHECK(sub_sum_of_squares(a, b, size(a) / 2, size(a)) == 0);
+    CHECK(sub_sum_of_squares(a, c, size(a) / 2, size(a)) == 5);
+    CHECK(sub_sum_of_squares(c, a, size(a) / 2, size(a)) == 5);
+    CHECK(sub_sum_of_squares(a, d, size(a) / 2, size(a)) == 5);
+    CHECK(sub_sum_of_squares(d, a, size(a) / 2, size(a)) == 5);
+    CHECK(sub_sum_of_squares(a, z, size(a) / 2, size(a)) == 730);
+  }
+
+  SECTION("sub_sum_of_squares_distance{} function object") {
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(a, b) == 0);
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(a, c) == 10);
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(c, a) == 10);
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(a, d) == 10);
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(d, a) == 10);
+    CHECK(sub_sum_of_squares_distance{0, size(a)}(a, z) == 985);
+
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(a, b) == 0);
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(a, c) == 5);
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(c, a) == 5);
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(a, d) == 5);
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(d, a) == 5);
+    CHECK(sub_sum_of_squares_distance{0, size(a) / 2}(a, z) == 255);
+
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(a, b) == 0);
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(a, c) == 5);
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(c, a) == 5);
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(a, d) == 5);
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(d, a) == 5);
+    CHECK(sub_sum_of_squares_distance{size(a) / 2, size(a)}(a, z) == 730);
+  }
+}
+
```

Comment:
@lums658 do we need this?

Reply:
Yes.  These function objects are what allow PQ to work with current query functions.